### DISCUSSION TITLE
Infinite loading after setting up directus locally

### DISCUSSION THREAD

**Discussion by antick** [directus/directus#8934] [Archive]
This is one of the best tools that I have seen after spending so much time researching something similar like this.

So I got it running through CLI and also the docker image. But there are few changes in the code that I would want to do locally so I followed this guide to set it up in my local env:

https://docs.directus.io/contributing/running-locally/

I have Postgresql installed in my local and I followed all these steps. I also ran 'npm run build'. Now the problem that I face is when I run `npm run dev` it randomly works. At one time it would open up the UI and on another, it would go into an infinite loading state with 0 warnings and errors in the console and network tab. 

If somehow it loads the UI then I am able to navigate through everything easily but if I hit the refresh button on the page then it would again go into the infinite loading state. It takes multiple tries to start it again and actually get it working.

I'm really clueless and not sure what exactly is going on here. Please help me out if anyone can understand this problem.

Thanks a lot!

### Update 1:

1. My database configuration is correct because if it's connected randomly then I am able to add the data but if I hit refresh it would get stuck again.
2. I have been trying the mentioned scenario with 👇
 npm run dev -- --scope directus --scope @directus/app
3. I'm running this on an Ubuntu 21.04 SSD Laptop with i5 and 16Gb RAM.
4. All the other node processes are closed and this is the only thing that I am running right now.

### Update 2:

Just figured out that a lot of requests are in a pending state in the network tab when I hit the admin URL. I waited for more than 3 hours and still found these requests in pending state in the network tab.

**Comment by azrikahar** [comment]
Hmm that does sound pretty odd. What is the OS of your local machine? Can you check _before_ you run `npm run dev`, are there any existing/unterminated node processes running in the background?

**Reply by antick** [reply]
Yes, I made sure that there are no other node processes running. I'm doing this on Ubuntu 21.04.

**Comment by joselcvarela** [comment]
Hello @antick ,
Check if your `.env` file has database configurations correct.
Also,  you need to run `npm install` and then `npm run build` first.
Then, instead of running `npm run dev` which runs every package, try to run only api and the app with
`npm run dev -- --scope directus --scope @directus/app`

**Reply by antick** [reply]
Hey, @joselcvarela 
Thanks for the reply.

Yes, the database config is correct. As I mentioned in my OP that I get it working few times randomly and I am able to add data into it once it's loaded. But if I hit the refresh then it's again stuck in the infinite loading or if I try to open it in the new tab it would go into the infinite connecting state with zero errors and no logs in the console.

If I turn the dev server off and run it again, I still see the same behavior. If I wait for some time and try again then it works for that time only. Hitting the refresh on the page would mess it up again.

I tried the above scenario with the below command only as I only start the app and not the packages:

`npm run dev -- --scope directus --scope @directus/app`

But unfortunately, it made no difference. I still see the same behavior that I mentioned above.

**Reply by joselcvarela** [reply]
I just remember I have a similar issue.
Try this:
- on one terminal run this: `npm run dev -- --scope directus` and wait until you see this: `✨ Server started at http://localhost:8055`
- once you see that message, open a new terminal and run `npm run dev -- --scope @directus/app`

By the way, where are you running?
I had to do some tricks to run it in WSL on Windows.

**Reply by antick** [reply]
Hey, just figured out that a lot of requests in the network tab are in a pending state.
And I am running it on an Ubuntu 21.04 laptop.

I tried the steps that you suggested. Both of those start okay but when I hit the admin URL in the browser, I can see the hundreds of requests in pending state.

**Reply by joselcvarela** [reply]
That's perfectly normal, it takes a while to start the app while in dev mode.
For example, it took me 38 seconds to start. That's because we have a lot of components on the app.

If it still do not work try this: `npm run dev -- --scope @directus/app -- -- --host=0.0.0.0`

**Reply by antick** [reply]
Usually, it takes less than 5 seconds for me when it randomly works 1 out of 10 times. But mostly it's stuck like this.

I checked it after 3 hours and still found it in the same pending state.

Also tried running it with `host=0.0.0.0` but still the same problem. I still see most of the requests in pending state.

**Reply by joselcvarela** [reply]
It seems the problem is with Vite and Chromium based browsers:
https://stackoverflow.com/questions/69481783/http-requests-stuck-in-pending-state-when-using-vite

Can you test with Firefox?

**Reply by antick** [reply]
Yes! It worked in Firefox!

Here is the bug that has been logged on vite repo for the same issue:

https://github.com/vitejs/vite/issues/5310#issuecomment-946899678

So this has nothing to do with directus directly.

Thanks a lot, @joselcvarela and @rijkvanzanten for helping me out here. I will follow up on that bug for the same.

You guys rock!

**Reply by joselcvarela** [reply]
Glad to be helpful ;)

**Comment by rijkvanzanten** [comment]
> Now the problem that I face is when I run npm run dev it randomly works. At one time it would open up the UI and on another, it would go into an infinite loading state with 0 warnings and errors in the console and network tab.

I _think_ i've seen this before. The issue might be a timing problem, where roughly the following happens:

The API starts up
While the API is starting up, a sub-package starts up
This sub-package change causes the API to restart
ts-node-dev gets confused, and doesn't ever start up anymore

this causes everything but the API to run, which in turn means the app can't reach the API.

I personally use what @joselcvarela suggested above. On first setup, run `npm run build` once to make sure all dependencies are available, then use `npm run dev` with the `-- --scope` option to select what packages to run (or use NPM workspaces: `npm run dev -w @directus/app` )

**Reply by antick** [reply]
Yes, I tried this as well where I only start the app and not the other packages. But unfortunately, this made no difference. 😥

**Reply by rijkvanzanten** [reply]
Very odd! Are there any warnings/errors printed in the browsers console log, or the terminal where you're running dev? 🤔

**Reply by antick** [reply]
No errors and warnings. It's blank. It's like it waits for something but not getting anything and is stuck in an endless loop. Nothing is printing on terminal and browser console.

If I keep it idle for some time and then come back and start the server, then it would work. But if I hit the refresh button it would be stuck again. Not sure what exactly is the problem here.

Do you think it has something to do with the build? I'm making sure that I run `npm run build` before I start the dev server.

**Reply by rijkvanzanten** [reply]
Hard to tell! If you check the network tab, is it timing out while fetching the app files? We're relying on `vite` for the app build, which fetches and compiles every file on the fly, so maybe the huge amount of compilations is slowing down your env to a point where the browser times out the request?

**Reply by antick** [reply]
Yes, you are right. I even had to increase the watcher limit otherwise I couldn't even come this far. This definitely has something to do with that because I checked the network tab requests and a lot of them are in a pending state for like forever.

**Comment by antick** [comment]
As mentioned by @joselcvarela that it is a known bug with `vite` and already reported here on vite repo a few days back: https://github.com/vitejs/vite/issues/5310#issuecomment-946899678

@joselcvarela comment: https://github.com/directus/directus/discussions/8934#discussioncomment-1503608


Currently, this issue is only reproducible on a Linux Chromium-based browser. It works fine on Firefox.

So this issue has nothing to do with directus.

Thanks a lot, @joselcvarela and @rijkvanzanten for helping me out here.

I'm putting this as an answer, in case anybody else faces the same issue.
